extensions/cache/deployment/src/test/java/io/quarkus/cache/test/runtime/SpecificKeysAndValuesCacheTest.java

```diff
@@ -0,0 +1,115 @@
+package io.quarkus.cache.test.runtime;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import javax.enterprise.context.Dependent;
+import javax.inject.Inject;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.cache.CacheLoad;
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ * Tests specific cache keys and values.
+ */
+public class SpecificKeysAndValuesCacheTest {
```

Comment:
I don't think we should support null keys nor null values. Although the JDK has collections that support them, I'm not aware of any major caching providers that support them because they are problematic.

Reply:
@galderz: This was introduced in this extension because I tried to store a `null` value and got yelled at by Caffeine, and then I saw in [Spring Cache](https://github.com/spring-projects/spring-framework/blob/master/spring-context/src/main/java/org/springframework/cache/support/AbstractValueAdaptingCache.java#L103) that `null` values were supported for all providers using a similar mechanism.

Reply:
I'd leave handling null values out for the time being. One extra knob to do what? It all boils down to what use case you're trying to solve that requires you to handle null values.

We have been developing Infinispan for over 10 years and we've never had a case where we missed not handling null values. JCache does not support null values either. Spring Cache might be the only one I'm aware of that handle null values.

Reply:
Ehcache 2.x does support [caching null values](https://www.ehcache.org/documentation/2.8/recipes/cachenull.html). That was a side effect of exposing their internal entry type (`Element`) and I suppose they featurized a bug. EOL though and 3.x doesn't support it. It is a lot less surprising to instead store an `Optional<V>` to mimic the behavior.

Reply:
My opinion is that if we are working on a general purpose extension which stores return values, we should support null. It's a pretty common case and making it harder for the user won't help our case.

Reply:
@gsmet Why we should support storing null? Could you explain this common case?

For ref, [here](https://groups.google.com/forum/#!topic/jsr107/OUXm9LAj8Ig) is a brief 2011 JSR-107 discussion on supporting null values.

Reply:
I was thinking of a method potentially returning null. It doesn't seem far fetched to me.

And sure, people could use Optional but if we can make things easier for the user without cluttering our code too much, I would do it.

Reply:
Here's my proposal about this discussion. We could remove the `null` support for now and consider reintroducing it later if there's a real demand from the Quarkus users. This PR is supposed to be a first iteration of the caching extension, it makes sense to exclude features that do not seem essential from it. WDYT?

Reply:
@galderz I've removed the null keys and values support from this PR, they are now both forbidden (a NPE will be thrown when either of them is used). They could also be silently ignored if you prefer that behaviour. I've also updated the Javadoc and extension guide accordingly. I'll keep the previous version of the code (which supported null keys and values) on a branch just in case we'd like to reintroduce this later.

Reply:
Null return value can be useful if not stored, like in Map computations. If the entry was not found, for instance.

Reply:
@ben-manes: Do you mean a null value returned by a method annotated with `@CacheLoad` or `@CacheStore` should be allowed (but not stored in the cache) and not be the cause of a `NPE` throw? The current version of the code still makes it possible to return null values as long as they're wrapped inside an `Optional`.

Reply:
An Optional is useful for negative caching where you want to cache the absence of the entry. A null could be useful to indicate absence but not cache. The various map computation methods let this be more powerful, e.g. removal. The flexibility might not be as useful in an annotation-based scheme, so just an FYI to consider.

Reply:
Reviewing this: I think I got a bit confused with this particular aspect of null values. Let's look at it case by case:

As suggested here, `CacheLoad` methods should be allowed to return null. This could happen if the underlying data source or service did not locate the item. This is fine, underneath we'd not cache anything. An optional can be returned, which as @ben-manes pointed out, can be used to cache the absence of something.

However, for `CacheStore` to return null would be odd. Remember that this annotation signals that the method, that might be inserting or updating a value in the data source or service, is storing a null value. We could allow it, but from a cache perspective, we'd just skip caching such a thing.

Reply:
I'll update this PR tonight to allow null values again.